zdict/dictionaries/yahoo.py

```diff
@@ -312,6 +310,6 @@ def parse_verbose(self, data):
                 'fw-500' in cls and ret.append(('explain', s))
 
             elif name == 'ul':
-                for li in node.select('> li'):
+                for li in node.select('li'):
```

Comment:
I don't actually know the difference between `> li` and `li`.

Reply:
`> li` indicates the direct children only, no any grandchildren included.

e.g. given the following snippet
```html
<ul>
  <li id="foo">
    <a>
      <ul><li id="bar"></li></ul>
    </a>
  </li>
```

`ul > li` get `foo` only, but `ul li` get both `foo` and `bar`.
So I said it's less strict in our code.

Reply:
I thought `>` is pretty basic CSS selector, they don't do that anymore !?

Reply:
I just try changing it to `ul > li`, looks like working as it should. Probably it just need put the current node type at the beginning ?

Reply:
If you need the `>` to be relative to the element select is being called on, use `:scope > li`.

I've been searching around for issues people have had with the new Soup Sieve to try and help the transition.

Reply:
Some reference reading on CSS selector level 4 `:scope`: https://developer.mozilla.org/en-US/docs/Web/CSS/:scope.

Reply:
Ah, @facelessuser how about adding more info on this type of error?
e.g.
```
SyntaxError: Cannot start or end selector with '>'. For the case of start with '>', lease use ':scope > ...' instead.
```

Reply:
It's just a syntax error. In CSS, you can't start a selector with `>`. `>` is a relational combinator. It has to have something before it and after it: `div > p` etc.  The only exception is with `:has()` because the current selector counts as the relational target. So for `div > p`, we are saying `p` must be a direct child of `div`, and we select the `p`.  With `div:has(> p)` it is the same, but we select `div`.

So when we say `> li`, we don't have a starting point, and it is invalid.  Try it in a browser console:

```js
> var el = document.querySelector('html')
> el.querySelector('> body')
VM2688:1 Uncaught DOMException: Failed to execute 'querySelector' on 'Element': '> body' is not a valid selector.
    at <anonymous>:1:4
```

But with the CSS level 4 selector, they've added the `:scope` selector.  In a normal CSS file (if implemented) it will select `:root` or the `html` tag. This is because there is no context.  But with `document.querySelector`, `:scope` represents the element that the selector is applied to.

```js
> el.querySelector(':scope > body')
<body>...</body>
```

In this case `:scope` equals the earlier selected `html` element.

In your case, `:scope` will be the selected `ul`, not all `ul`s.  Just the selected one.

Hopefully that makes sense.

Reply:
yeah, I updated the code, it works fine now.

ah, so we abused the selector previously, right?

Reply:
The previous selector used a very naive algorithm. I don't know if allowing `> li` was intentional or not, but it definitely didn't follow the spec.  It apparently allowed things like `td.+.class` which is also wrong since `td` is followed by a `.` and no class name.

Soup Sieve tries to follow the CSS spec (to the best of my understanding and to the best that can be implemented), so it will throw a syntax error if you are doing something unrecognized or invalid.

Reply:
@facelessuser Thanks for the info!

Personally, I still think adding more suggest in error message helps a lot.

Probably something like this:

```
SyntaxError: Cannot start or end selector with '>'.

Search these for what you want:
 
* ':scope > ...'
* '{current_node_name} > ...'
* '... > ...'
```

Reply:
@facelessuser Pretty clear explanation. Thanks a lot!

Reply:
@facelessuser From my experience in [Rust](https://www.rust-lang.org/) (another programming language), more info and suggestion in error message helps **a lot**! They improve their compiler error messages over and over again, people even build tool to automatically apply the compiler suggestion.

Overall, I just want to share my experience about the important of error messages. The error might caused by users' misuse, but good better message will help user experience! 

FYI, most of `zdict` developers are not super familar with CSS. We all wrote some web before but that's not our main skill.

Reply:
@wdv4758h, I understand, and I did actually commit a change to update the messages, maybe not as verbose as you posted earlier, but I did try to be a little more specific: https://github.com/facelessuser/soupsieve/pull/72/files#diff-f0f319dd801dd20448eedbfd429fc221R586.

The main reason I was poking around in issues related to Soup Sieve was to get exactly this kind of feedback. So I did follow through with your suggestion to some extent.  If it turns out people still get tripped up on certain exceptions, I will most likely update again and be more verbose if required.

Though it should be mentioned that Soup Sieve does not try to provide complete documentation for CSS selectors, that is left up to sites like: https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Selectors. Or even the CSS spec: https://drafts.csswg.org/selectors-4/.  The Soup Sieve documentation, gives only brief examples with the expectation that people will reference more complete CSS documentation.  It is more informing what is available in Soup Sieve. Though it probably wouldn't hurt for me to link things like the CSS spec in the Soup Sieve documentation.